support/agent/miner_agent.go

```diff
@@ -0,0 +1,338 @@
+	if err != nil {
+		return err
+	}
+
+	if len(ma.deadlines[dlIdx]) == 0 {
+		err := ma.scheduleNextProof(v, dlIdx)
+		if err != nil {
+			return err
+		}
+	}
+
+	// pIdx should be sequential, but add empty partitions just in case
+	for pIdx >= uint64(len(ma.deadlines[dlIdx])) {
+		ma.deadlines[dlIdx] = append(ma.deadlines[dlIdx], bitfield.New())
+	}
+	ma.deadlines[dlIdx][pIdx].Set(uint64(sectorNumber))
+	return nil
+}
+
+// schedule a proof within the deadline's bounds
+func (ma *MinerAgent) scheduleNextProof(v *vm.VM, dlIdx uint64) error {
+	var st miner.State
+	err := v.GetState(ma.IDAddress, &st)
+	if err != nil {
+		return err
+	}
+
+	// find next proving window for this deadline
+	deadlineStart := st.ProvingPeriodStart + abi.ChainEpoch(dlIdx)*miner.WPoStChallengeWindow
+	if deadlineStart-miner.WPoStChallengeWindow < v.GetEpoch() {
+		deadlineStart += miner.WPoStProvingPeriod
+	}
+	deadlineClose := deadlineStart + miner.WPoStChallengeWindow
+	prooveAt := deadlineStart + abi.ChainEpoch(ma.rnd.Int63n(int64(deadlineClose-deadlineStart)))
+	ma.operationSchedule.ScheduleOp(prooveAt, proveDeadlineAction{dlIdx: dlIdx})
+	return nil
+}
+
+// create prove commit message
+func (ma *MinerAgent) createProveCommit(sectorNumber abi.SectorNumber) Message {
+	params := miner.ProveCommitSectorParams{
+		SectorNumber: sectorNumber,
+	}
+
+	return Message{
+		From:          ma.Worker,
+		To:            ma.RobustAddress,
+		Value:         big.Zero(),
+		Method:        builtin.MethodsMiner.ProveCommitSector,
+		Params:        &params,
+		ReturnHandler: ma.handleProveCommit,
+	}
+}
+
+// register an op for next epoch (after batch prove) to schedule a post for the sector
+func (ma *MinerAgent) handleProveCommit(v *vm.VM, msg Message, _ cbor.Marshaler) error {
+	sn := msg.Params.(*miner.ProveCommitSectorParams).SectorNumber
+	ma.PreCommits--
+	ma.operationSchedule.ScheduleOp(v.GetEpoch(), registerSectorAction{sectorNumber: sn})
+	return nil
+}
+
+// create precommit message and activation trigger
+func (ma *MinerAgent) createPrecommit(currentEpoch abi.ChainEpoch, sectorNumber abi.SectorNumber, sectorActivation abi.ChainEpoch) Message {
+	params := miner.PreCommitSectorParams{
+		SealProof:     ma.Config.ProofType,
+		SectorNumber:  sectorNumber,
+		SealedCID:     sectorSealCID(ma.rnd),
+		SealRandEpoch: currentEpoch - 1,
+		Expiration:    ma.sectorExpiration(currentEpoch),
+	}
+	return Message{
+		From:   ma.Worker,
+		To:     ma.RobustAddress,
+		Value:  big.Zero(),
+		Method: builtin.MethodsMiner.PreCommitSector,
+		Params: &params,
+		ReturnHandler: func(_ *vm.VM, _ Message, _ cbor.Marshaler) error {
+			ma.PreCommits++
+			ma.operationSchedule.ScheduleOp(sectorActivation, proveCommitAction{sectorNumber})
+			return nil
+		},
+	}
+}
+
+// create a random valid sector expiration
+func (ma *MinerAgent) sectorExpiration(currentEpoch abi.ChainEpoch) abi.ChainEpoch {
+	// these are precommit bounds. Prove commit is more lenient but contains this range.
```

Comment:
I made similar assumptions, but I kept hitting errors. Here are the constraints on expiration:

**PreCommit**
1. [activation := precommitEpoch + MaxProveCommitDuration](https://github.com/filecoin-project/specs-actors/blob/74047a94726e66d895b4fb672f050f61b4360e8b/actors/builtin/miner/miner_actor.go#L555)
2. [expiration > activation](https://github.com/filecoin-project/specs-actors/blob/74047a94726e66d895b4fb672f050f61b4360e8b/actors/builtin/miner/miner_actor.go#L555)
3. [expiration >= MinSectorExpiration + activation](https://github.com/filecoin-project/specs-actors/blob/74047a94726e66d895b4fb672f050f61b4360e8b/actors/builtin/miner/miner_actor.go#L2002) (note this is stronger than 2.)
4. [expiration <= rt.CurrEpoch()+MaxSectorExpirationExtension](https://github.com/filecoin-project/specs-actors/blob/74047a94726e66d895b4fb672f050f61b4360e8b/actors/builtin/miner/miner_actor.go#L2008)
5. [expiration <= builtin.SealProofSectorMaximumLifetime(sealProof) + activation](https://github.com/filecoin-project/specs-actors/blob/74047a94726e66d895b4fb672f050f61b4360e8b/actors/builtin/miner/miner_actor.go#L2016)

**ProveCommit**
Does not validate expiration

**ConfirmSectorProofsValid**
1. [expiration >= rt.CurrEpoch(prove) + MinSectorExpiration](https://github.com/filecoin-project/specs-actors/blob/74047a94726e66d895b4fb672f050f61b4360e8b/actors/builtin/miner/miner_actor.go#L865) (skips sector but does not error)

Since rt.CurrEpoch() + MaxProveCommitDuration[params.SealProof] at precommit time is greater than or equal rt.CurrEpoch() at prove time, PreCommit (3) is a stronger constraint than ConfirmSectorProofsValid (1). I might have made an error in this analysis. I vaguely remember setting the ranges like this on purpose to avoid potential errors in ConfirmSectorProofsValid.

Reply:
This would benefit from explanatory comments [like this](https://github.com/filecoin-project/specs-actors/blob/74047a94726e66d895b4fb672f050f61b4360e8b/actors/builtin/miner/miner_actor.go#L553). Note that since expiration is validated at pre-commit, it conservatively assumes the latest possible activation, and then requires at least `MinSectorExpiration` additional lifetime from that point.